---
### Question 1: [Index] S&P 500 Stocks Added to the Index

**Which year had the highest number of additions?**

Using the list of S&P 500 companies from Wikipedia's [S&P 500 companies page](https://en.wikipedia.org/wiki/List_of_S%26P_500_companies), download the data including the year each company was added to the index.

Hint: you can use [pandas.read_html](https://pandas.pydata.org/docs/reference/api/pandas.read_html.html) to scrape the data into a DataFrame.

Steps:
1. Create a DataFrame with company tickers, names, and the year they were added.
2. Extract the year from the addition date and calculate the number of stocks added each year.
3. Which year had the highest number of additions (1957 doesn't count, as it was the year when the S&P 500 index was founded)? Write down this year as your answer (the most recent one, if you have several records).

*Context*: 
> "Following the announcement, all four new entrants saw their stock prices rise in extended trading on Friday" - recent examples of S&P 500 additions include DASH, WSM, EXE, TKO in 2025 ([Nasdaq article](https://www.nasdaq.com/articles/sp-500-reshuffle-dash-tko-expe-wsm-join-worth-buying)).

*Additional*: How many current S&P 500 stocks have been in the index for more than 20 years? When stocks are added to the S&P 500, they usually experience a price bump as investors and index funds buy shares following the announcement.

---

In [1]:
import pandas as pd


In [5]:
! pip install lxml

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 926.0 kB/s eta 0:00:00a 0:00:01


In [14]:
url = "https://en.wikipedia.org/wiki/List_of_S%26P_500_companies"
tables = pd.read_html(url)

In [15]:
type(tables), len(tables)

(list, 2)

In [16]:
tickers = tables[0]

In [17]:
tickers.head()

Symbol             Security             GICS Sector  \
0    MMM                   3M             Industrials   
1    AOS          A. O. Smith             Industrials   
2    ABT  Abbott Laboratories             Health Care   
3   ABBV               AbbVie             Health Care   
4    ACN            Accenture  Information Technology   

                GICS Sub-Industry    Headquarters Location  Date added  \
0        Industrial Conglomerates    Saint Paul, Minnesota  1957-03-04   
1               Building Products     Milwaukee, Wisconsin  2017-07-26   
2           Health Care Equipment  North Chicago, Illinois  1957-03-04   
3                   Biotechnology  North Chicago, Illinois  2012-12-31   
4  IT Consulting & Other Services          Dublin, Ireland  2011-07-06   

       CIK      Founded  
0    66740         1902  
1    91142         1916  
2     1800         1888  
3  1551152  2013 (1888)  
4  1467373         1989

In [18]:
tickers.dtypes

Symbol                   object
Security                 object
GICS Sector              object
GICS Sub-Industry        object
Headquarters Location    object
Date added               object
CIK                       int64
Founded                  object
dtype: object

In [ ]:
tickers["Date added"]

0      1957-03-04
1      2017-07-26
2      1957-03-04
3      2012-12-31
4      2011-07-06
          ...    
498    2011-11-01
499    1997-10-06
500    2019-12-23
501    2001-08-07
502    2013-06-21
Name: Date added, Length: 503, dtype: object

In [21]:
tickers.columns = tickers.columns.str.lower().str.replace(' ', '_')

In [22]:
tickers.head()

symbol             security             gics_sector  \
0    MMM                   3M             Industrials   
1    AOS          A. O. Smith             Industrials   
2    ABT  Abbott Laboratories             Health Care   
3   ABBV               AbbVie             Health Care   
4    ACN            Accenture  Information Technology   

                gics_sub-industry    headquarters_location  date_added  \
0        Industrial Conglomerates    Saint Paul, Minnesota  1957-03-04   
1               Building Products     Milwaukee, Wisconsin  2017-07-26   
2           Health Care Equipment  North Chicago, Illinois  1957-03-04   
3                   Biotechnology  North Chicago, Illinois  2012-12-31   
4  IT Consulting & Other Services          Dublin, Ireland  2011-07-06   

       cik      founded  
0    66740         1902  
1    91142         1916  
2     1800         1888  
3  1551152  2013 (1888)  
4  1467373         1989

In [24]:
tickers['date_added'] = pd.to_datetime(tickers['date_added'])

In [25]:
tickers.head()

symbol             security             gics_sector  \
0    MMM                   3M             Industrials   
1    AOS          A. O. Smith             Industrials   
2    ABT  Abbott Laboratories             Health Care   
3   ABBV               AbbVie             Health Care   
4    ACN            Accenture  Information Technology   

                gics_sub-industry    headquarters_location date_added  \
0        Industrial Conglomerates    Saint Paul, Minnesota 1957-03-04   
1               Building Products     Milwaukee, Wisconsin 2017-07-26   
2           Health Care Equipment  North Chicago, Illinois 1957-03-04   
3                   Biotechnology  North Chicago, Illinois 2012-12-31   
4  IT Consulting & Other Services          Dublin, Ireland 2011-07-06   

       cik      founded  
0    66740         1902  
1    91142         1916  
2     1800         1888  
3  1551152  2013 (1888)  
4  1467373         1989

In [27]:
tickers.dtypes

symbol                           object
security                         object
gics_sector                      object
gics_sub-industry                object
headquarters_location            object
date_added               datetime64[ns]
cik                               int64
founded                          object
dtype: object

In [28]:
tickers['year'] = tickers['date_added'].dt.year


In [29]:
tickers.head()

symbol             security             gics_sector  \
0    MMM                   3M             Industrials   
1    AOS          A. O. Smith             Industrials   
2    ABT  Abbott Laboratories             Health Care   
3   ABBV               AbbVie             Health Care   
4    ACN            Accenture  Information Technology   

                gics_sub-industry    headquarters_location date_added  \
0        Industrial Conglomerates    Saint Paul, Minnesota 1957-03-04   
1               Building Products     Milwaukee, Wisconsin 2017-07-26   
2           Health Care Equipment  North Chicago, Illinois 1957-03-04   
3                   Biotechnology  North Chicago, Illinois 2012-12-31   
4  IT Consulting & Other Services          Dublin, Ireland 2011-07-06   

       cik      founded  year  
0    66740         1902  1957  
1    91142         1916  2017  
2     1800         1888  1957  
3  1551152  2013 (1888)  2012  
4  1467373         1989  2011

In [30]:
import datetime

tickers['age'] = datetime.datetime.now().year - tickers['date_added'].dt.year


In [31]:
tickers.head()

symbol             security             gics_sector  \
0    MMM                   3M             Industrials   
1    AOS          A. O. Smith             Industrials   
2    ABT  Abbott Laboratories             Health Care   
3   ABBV               AbbVie             Health Care   
4    ACN            Accenture  Information Technology   

                gics_sub-industry    headquarters_location date_added  \
0        Industrial Conglomerates    Saint Paul, Minnesota 1957-03-04   
1               Building Products     Milwaukee, Wisconsin 2017-07-26   
2           Health Care Equipment  North Chicago, Illinois 1957-03-04   
3                   Biotechnology  North Chicago, Illinois 2012-12-31   
4  IT Consulting & Other Services          Dublin, Ireland 2011-07-06   

       cik      founded  year  age  
0    66740         1902  1957   68  
1    91142         1916  2017    8  
2     1800         1888  1957   68  
3  1551152  2013 (1888)  2012   13  
4  1467373         1989  2011   14

In [34]:
tickers['year'].value_counts().sort_values(ascending=False).head(5)

year
1957    53
2017    23
2016    23
2019    22
2008    17
Name: count, dtype: int64

In [37]:
tickers[tickers.age > 20].sort_values(ascending=False, by="age")

symbol                  security             gics_sector  \
0      MMM                        3M             Industrials   
2      ABT       Abbott Laboratories             Health Care   
25     AEP   American Electric Power               Utilities   
21      MO                    Altria        Consumer Staples   
43     ADM    Archer Daniels Midland        Consumer Staples   
..     ...                       ...                     ...   
281     LH                   Labcorp             Health Care   
297    MTB                  M&T Bank              Financials   
434    TPR            Tapestry, Inc.  Consumer Discretionary   
444    TMO  Thermo Fisher Scientific             Health Care   
472   VTRS                   Viatris             Health Care   

                       gics_sub-industry       headquarters_location  \
0               Industrial Conglomerates       Saint Paul, Minnesota   
2                  Health Care Equipment     North Chicago, Illinois   
25                    Electric Utilities              Columbus, Ohio   
21                               Tobacco          Richmond, Virginia   
43      Agricultural Products & Services           Chicago, Illinois   
..                                   ...                         ...   
281                 Health Care Services  Burlington, North Carolina   
297                       Regional Banks           Buffalo, New York   
434  Apparel, Accessories & Luxury Goods     New York City, New York   
444       Life Sciences Tools & Services      Waltham, Massachusetts   
472                      Pharmaceuticals    Pittsburgh, Pennsylvania   

    date_added      cik      founded  year  age  
0   1957-03-04    66740         1902  1957   68  
2   1957-03-04     1800         1888  1957   68  
25  1957-03-04     4904         1906  1957   68  
21  1957-03-04   764180         1985  1957   68  
43  1957-03-04     7084         1902  1957   68  
..         ...      ...          ...   ...  ...  
281 2004-11-01   920148         1978  2004   21  
297 2004-02-23    36270         1856  2004   21  
434 2004-09-01  1116132         2017  2004   21  
444 2004-08-03    97745  2006 (1902)  2004   21  
472 2004-04-23  1792044         1961  2004   21  

[219 rows x 10 columns]

In [38]:
tickers[tickers.age > 20].shape[0]

219

## Question 1

* Which year had the highest number of additions (1957 doesn't count, as it was the year when the S&P 500 index was founded)? Write down this year as your answer (the most recent one, if you have several records).

* Additional: How many current S&P 500 stocks have been in the index for more than 20 years? When stocks are added to the S&P 500, they usually experience a price bump as investors and index funds buy shares following the announcement.

## Answer 1


* 2017
* Additional = 219